<a href="https://colab.research.google.com/github/annakysilova/Data-Fr/blob/main/LabcsvKysilovaAnna.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Устанавливаем через pip необходимые пакеты**

In [1]:
!pip install psycopg2-binary

     |████████████████████████████████| 3.0MB 7.0MB/s 


**Подключение к БД PostgreSQL на основе sqlalchemy**

In [2]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sqlalchemy import create_engine

#Строка подключения - из Heroku
postgres_str='postgres://lrudzgbepzxxig:16401e013e429e971afef55ced20620c96f7ac914c9e98a135983616f2b40b50@ec2-3-220-23-212.compute-1.amazonaws.com:5432/d4i2dneufab066'

#Создание курсора для работы с БД
cnx = create_engine(postgres_str)

#Запрос на получение всех таблиц из БД
pd.read_sql_query('SELECT * FROM pg_catalog.pg_tables', cnx)

schemaname                tablename  ... hastriggers rowsecurity
0               public                playstore  ...       False       False
1           pg_catalog             pg_statistic  ...       False       False
2           pg_catalog                  pg_type  ...       False       False
3           pg_catalog        pg_foreign_server  ...       False       False
4           pg_catalog                pg_authid  ...       False       False
..                 ...                      ...  ...         ...         ...
66  information_schema  sql_implementation_info  ...       False       False
67  information_schema                sql_parts  ...       False       False
68  information_schema            sql_languages  ...       False       False
69  information_schema               sql_sizing  ...       False       False
70  information_schema      sql_sizing_profiles  ...       False       False

[71 rows x 8 columns]

**Загрузка данных из CSV-файла**

In [3]:
df = pd.read_csv("drive/MyDrive/Data Science/cvs/googleplaystore1.csv")

**Вывод первых 10 строк**

In [ ]:
df.head()

**Загрузим датафрейм с даныыми в БД в таблицу playstore**

In [ ]:
df.to_sql("playstore", con=cnx)

**Сделаем запрос на выборку из БД**

In [ ]:
q=pd.read_sql_query('SELECT * FROM playstore;', cnx)

q

**Посмотрим структуру полей БД средствами Pandas**

In [ ]:
q.columns

**В качестве примера рассмотрим гипотезу, что рейтинг зависит от количества инсталляций. Для подтверждения этой гипотезы построим регрессионную модель Rating от Installs**

In [9]:
print(q.Rating, q.Installs)

0       4.1
1       3.9
2       4.7
3       4.5
4       4.3
       ... 
5812    NaN
5813    3.7
5814    3.9
5815    4.8
5816    NaN
Name: Rating, Length: 5817, dtype: float64 0           10,000+
1          500,000+
2        5,000,000+
3       50,000,000+
4          100,000+
           ...     
5812             1+
5813       100,000+
5814        10,000+
5815         1,000+
5816         1,000+
Name: Installs, Length: 5817, dtype: object


**Анализ данных показывает, что Installs является текстовым полем, которое содержит разделители "," и индикатор "+" в конце строки. Необходимо преобразовать значения из строковых в числовые**


In [10]:
#Убрали последний символ в строке
q.Installs= q.Installs.map(lambda x: str(x)[:-1])

In [11]:
q.Installs

0           10,000
1          500,000
2        5,000,000
3       50,000,000
4          100,000
           ...    
5812             1
5813       100,000
5814        10,000
5815         1,000
5816         1,000
Name: Installs, Length: 5817, dtype: object

In [12]:
#Убрали ',' в строке
q.Installs = [x.replace(',', '') for x in q.Installs]

In [13]:
q.Installs

0          10000
1         500000
2        5000000
3       50000000
4         100000
          ...   
5812           1
5813      100000
5814       10000
5815        1000
5816        1000
Name: Installs, Length: 5817, dtype: object

**Уберем пустые значение в полях Installs и Rating**

In [14]:
q[q.Installs=='']

Empty DataFrame
Columns: [index, App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []

In [15]:
#Удаление пустых значений
q=q.drop(q[q.Installs==None].index)

In [ ]:
#Удаление всех записей с пустыми значениями
q=q.dropna()
q

**Меняем тип поля с строки в float**

In [17]:
#!pip install pandas

In [18]:
q.Installs=q.Installs.astype(float)

In [19]:
q.Installs

0          10000.0
1         500000.0
2        5000000.0
3       50000000.0
4         100000.0
           ...    
5808       10000.0
5811        1000.0
5813      100000.0
5814       10000.0
5815        1000.0
Name: Installs, Length: 5456, dtype: float64

In [20]:
#Если нужно использовать float64
q.Installs= q.Installs.astype(np.float64)

In [21]:
#Удаляем поля, которые были не пустые, а содержали текст NaN
q=q.drop(q[q.Rating.isin(['NaN'])].index)

In [ ]:
# Выборка по условию Installs<100
q[q.Installs<100]

**Сделаем преобразование данных (нормализацию)**

In [23]:
#Будем использовать датафрейм q_s с нормализованными данными
q_s=q 

In [24]:
q_s.Installs=q_s.Installs.transform(lambda x: (x / x.max()))

In [25]:
q_s.Installs.max()

1.0

In [26]:
q_s.Installs

0       0.000010
1       0.000500
2       0.005000
3       0.050000
4       0.000100
          ...   
5808    0.000010
5811    0.000001
5813    0.000100
5814    0.000010
5815    0.000001
Name: Installs, Length: 5456, dtype: float64

In [27]:
q_s.Rating=q_s.Rating.transform(lambda x: (x / x.max()))

In [28]:
q_s.Rating

0       0.82
1       0.78
2       0.94
3       0.90
4       0.86
        ... 
5808    0.60
5811    0.74
5813    0.74
5814    0.78
5815    0.96
Name: Rating, Length: 5456, dtype: float64

In [ ]:
#  Rating - 3 поле, Installs - 6 поле
q_s.columns

**Создаем функцию для построения регрессионной модели на основе sklearn.linear_model**

In [30]:
q.columns

Index(['index', 'App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs',
       'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated',
       'Current Ver', 'Android Ver'],
      dtype='object')

In [31]:
import pickle
def model_(data_model):
   
    Y = data_model.iloc[:,[4]] # Rating - 3 поле
    X=data_model.iloc[:,[6]]#Installs -6 поле
    
    filename = 'Model.sav'
    
    model = LinearRegression().fit(X,Y) #Обучение модели
    r_sq = model.score(X,Y) #Оценка качества модели
    print(r_sq)
    print(model.predict(X[0:1]))#Прогнозирование для 1-ого наблюдения
    pickle.dump(model, open(filename, 'wb')) #Сохранение модели на диске
    
    #loaded_model = pickle.load(open(str(filename), 'rb'))#Загрузка модели 
    
    #forecast_score = loaded_model.predict(X[0:1])#Использование загруженной модели 

**Моделируем**

In [32]:
model_(q)

0.002561364423799506
[[0.84712009]]


**Результаты моделирования (R^2 = 0.0026) показывают осутствие зависимости между Rating и Installs**

In [ ]:
model_(q_s) #Это справедливо и для нормализованных данных

## ЗАДАНИЕ:
**Подтвердить гипотезу о зависимости между Reviews и Installs для приложений, размер которых меньше 20M**

**Проведем нормализацию для Reviews**

In [35]:
q_r=q

In [36]:
q_r.Installs=q_r.Installs.transform(lambda x: (x / x.max()))

In [37]:
q_r.Installs.max()

1.0

In [38]:
q_r.Installs

0       0.000010
1       0.000500
2       0.005000
3       0.050000
4       0.000100
          ...   
5808    0.000010
5811    0.000001
5813    0.000100
5814    0.000010
5815    0.000001
Name: Installs, Length: 5456, dtype: float64

In [39]:
q_r.Reviews=q_r.Reviews.transform(lambda x: (x / x.max()))

In [40]:
q_r.Rating

0       0.82
1       0.78
2       0.94
3       0.90
4       0.86
        ... 
5808    0.60
5811    0.74
5813    0.74
5814    0.78
5815    0.96
Name: Rating, Length: 5456, dtype: float64

In [41]:
q_r.columns #Вывод наших нормализированных Reviews и Installs 

Index(['index', 'App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs',
       'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated',
       'Current Ver', 'Android Ver'],
      dtype='object')

**Создаем функцию для построения регрессионной модели на основе sklearn.linear_model**

In [42]:
q.columns

Index(['index', 'App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs',
       'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated',
       'Current Ver', 'Android Ver'],
      dtype='object')

In [49]:
import pickle
def model_(data_model):
   
    Y = data_model.iloc[:,[4]] # Reviews - 4 поле
    X=data_model.iloc[:,[6]] #Installs -6 поле
    
    filename = 'Model.sav'
    
    model = LinearRegression().fit(X,Y) #Обучение модели
    r_sq = model.score(X,Y) #Оценка качества модели
    print(r_sq)
    print(model.predict(X[0:1]))#Прогнозирование для 1-ого наблюдения
    pickle.dump(model, open(filename, 'wb')) #Сохранение модели на диске
    
    #loaded_model = pickle.load(open(str(filename), 'rb'))#Загрузка модели 
    
    #forecast_score = loaded_model.predict(X[0:1])#Использование загруженной модели 

Моделируем

In [50]:
model_(q)

0.4096719508027953
[[0.00228229]]


Результаты моделирования (R^2 = 0.4097) показывают осутствие зависимости между Rating и Installs


In [51]:
model_(q_r) #Это справедливо и для нормализованных данных

0.4096719508027953
[[0.00228229]]
